### PR TITLE
#4123: example state allow for continue and restart

### LINKED ISSUE #4123: Improve state change example
## What problem does this solve or what need does it fill?

In the example for changing [state](https://github.com/bevyengine/bevy/blob/latest/examples/ecs/state.rs), it shows a great example how to which from state Menu to state InGame. However it is missing, how to deal with going back from the game, to the menu. 

## What solution would you like?

I would propose an addition to the example, where pressing the escape key moves you back to the menu. Once back in the menu you've got two option:

- Continue the game where the player left off (meaning reentering the stage "InGame" with the already existing entities)
- Create a new game where it clears out the existing entities (like the first time loading the game)

## What alternative(s) have you considered?

The documentation on "stages" and/or "systemset" could be improved to explain a bit more what happens and is expected when SystemSet::on_enter(), SystemSet::on_update(), etc are triggered

### PR DESCRIPTION
**Pull request by Vrixyz**
# Objective

Fixes #4123.

## Solution

- menu state is now more dynamic, with 2 scenario:
  - first load: displaying **Play**, and setting up camera 🤷 
  - subsesquent loads: **Restart** and **Continue**
- It might be cleaner to use more states (Loading and/or first menu VS PauseMenu) ? But it would add more complexity so I tried to reuse code as much as possible.

States are now:
- (first)Menu
- Game
- Game -> (**pushed** subsequent) Menu
  - `Restart` leads to a replace of the whole State stack to Game (calling `on_exit` and `on_enter`)
  - `Continue`simply pops the last menu state.

We could make this example better by avoiding the blue changing color to quickly change (use delta_time rather than time), to add to the feeling of real "pause" and start from previous state.

### CODE REVIEW

File: examples/ecs/state.rs [hunk 1]
```diff
@@ -27,16 +29,44 @@ enum AppState {
...
+        continue_entity = Some(create_button(&mut commands, &*asset_server, "Continue"));
+    }
+
+    commands.insert_resource(MenuData {
+        button_entity,
+        continue_entity,
+    });
+}
+
+fn create_button(commands: &mut Commands, asset_server: &AssetServer, title: &str) -> Entity {
```
Comment (alice-i-cecile) [inline examples/ecs/state.rs:68]: I like this; users often struggle to figure out how to write simple abstractions like this for manipulating entities.

File: examples/ecs/state.rs [hunk 2]
```diff
@@ -27,16 +29,44 @@ enum AppState {
...
-fn setup_menu(mut commands: Commands, asset_server: Res<AssetServer>) {
-    // ui camera
-    commands.spawn_bundle(UiCameraBundle::default());
-    let button_entity = commands
+fn setup_menu(
+    mut state: ResMut<State<AppState>>,
+    mut commands: Commands,
+    asset_server: Res<AssetServer>,
+) {
+    let is_first_load = !state.inactives().contains(&AppState::InGame);
```
[the remark this answers is not in the document]
Comment (Vrixyz) [reply examples/ecs/state.rs:47]: I hesitated to modify Menu state into Menu(bool) or inner clearer enum to carry this information.

File: examples/ecs/state.rs [hunk 3]
```diff
@@ -94,16 +134,34 @@ fn menu(
 
 fn cleanup_menu(mut commands: Commands, menu_data: Res<MenuData>) {
     commands.entity(menu_data.button_entity).despawn_recursive();
+    if menu_data.continue_entity.is_some() {
+        commands
+            .entity(menu_data.continue_entity.unwrap())
+            .despawn_recursive();
+    }
+}
+fn cleanup_game(mut commands: Commands, query_to_remove: Query<Entity, With<RemoveWhenGameDone>>) {
```
Comment (alice-i-cecile) [inline examples/ecs/state.rs:143]:
```suggestion

fn cleanup_game(mut commands: Commands, query_to_remove: Query<Entity, With<RemoveWhenGameDone>>) {
```

File: examples/ecs/state.rs [hunk 4]
```diff
@@ -94,16 +134,34 @@ fn menu(
...
+        .spawn_bundle(OrthographicCameraBundle::new_2d())
+        .insert(RemoveWhenGameDone);
+    commands
+        .spawn_bundle(SpriteBundle {
+            texture: asset_server.load("branding/icon.png"),
+            ..default()
+        })
+        .insert(RemoveWhenGameDone);
+}
+fn back_to_menu(time: Res<Time>, mut state: ResMut<State<AppState>>, input: Res<Input<KeyCode>>) {
```
Comment (alice-i-cecile) [inline examples/ecs/state.rs:160]:
```suggestion

fn back_to_menu(time: Res<Time>, mut state: ResMut<State<AppState>>, input: Res<Input<KeyCode>>) {
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I like this! I think that once we've removed the state stack stuff this will make a nice improvement.

File: examples/ecs/state.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply examples/ecs/state.rs:47]: I was actually just thinking a resource that we flip that encodes *why* we care about this information.

**Review by alice-i-cecile — APPROVED** [review]
This LGTM; is this ready to come out of draft mode?

**Comment by Vrixyz** [comment]
I'm closing this as it's way outdated now

**Pull request closed without merging**